Safe scatter

Redmar-van-den-Berg opened the pull request:
I added a new scatter method called `safe-scatter`, which guarantees a minimum size for all scattered regions, and has an almost perfectly even distribution across all bed files. 
It also has an option to mix the small regions in between the larger regions, which is relevant for the human reference genome since the small unplaced and random contigs take significantly longer to call then regular chromosomes. This effect will be even more pronounced with HG38, since it will have even more small contigs.

It looks like the last test failed because codecov is angry I removed a single unused import.
### Checklist
- [x] Pull request details were added to CHANGELOG.rst
- [x] Documentation updated (if required)

Redmar-van-den-Berg committed:
Begin adding safe-scatter

Redmar-van-den-Berg committed:
Add additional tests for scatter_size

Redmar-van-den-Berg committed:
Change terminology

Redmar-van-den-Berg committed:
Add function to scatter a region

Redmar-van-den-Berg committed:
Add support for regions smaller than min_scatter_size

Redmar-van-den-Berg committed:
Fix flake8 errors

Redmar-van-den-Berg committed:
Add test for invalid scatter sizes

Redmar-van-den-Berg committed:
Add safe_scatter function and tests

Redmar-van-den-Berg committed:
Add some test for valid and invalid edge cases

Redmar-van-den-Berg committed:
Fix some flake8 errors

Redmar-van-den-Berg committed:
Fix a mypy error

Redmar-van-den-Berg committed:
Add option to shuffle the input regions

Redmar-van-den-Berg committed:
Remove split-contig flag from safe-scatter

Redmar-van-den-Berg committed:
Update documentation and changelog

Redmar-van-den-Berg committed:
Fix flake8 error

Redmar-van-den-Berg committed:
Add function to mix small regions between large regions

Redmar-van-den-Berg committed:
Drop shuffling in favor of mixing small regions

Redmar-van-den-Berg committed:
Remove unused import

rhpvorderman commented on src/chunked_scatter/safe_scatter.py:
```diff
@@ -0,0 +1,244 @@
+
+def sum_regions(regions: List[BedRegion]):
+    """ Calculate the total length of all regions """
+    return sum(len(region) for region in regions)
```
This number will not be entirely correct when there are overlapping regions. Is that intentional?
`return sum(len(region) for region in merge_regions(regions))` is also an option.

Redmar-van-den-Berg replied:
There should be no overlapping regions, since the regions are generated based on the number of requested chunks. So it should not be possible to get overlapping regions into this function.

rhpvorderman commented on src/chunked_scatter/safe_scatter.py:
```diff
@@ -0,0 +1,244 @@
+        regions and end up with a whole bunch of uninterrupted regular regions.
+    """
+    # Small regions are regions that are smaller than the target_bin_size
+    regular_regions = [reg for reg in regions if len(reg) >= target_bin_size]
```
Have a look at https://docs.python.org/3/library/itertools.html#itertools-recipes

The partition example accomplishes the same thing. But you only have to write the condition once and the other list is automatically the inverse condition.

I don't know if that is necessary here. This looks readable as well. But if there are a lot of contigs the itertools example will be slightly more efficient (premature optimization :rocket:). This is very readable though. So maybe better to keep this.

Redmar-van-den-Berg replied:
Partition is quite a nice trick, but I don't think it is needed here. Even if you have a lot of contigs (a few thousand?) this code should still run fast. There is a big difference between 'a lot of contigs biologically', which is a few thousand, and 'a lot of contigs mathematically', which would be in the millions.

rhpvorderman replied:
Agreed. At least it is fun to know what tricks are out there.

rhpvorderman commented on src/chunked_scatter/safe_scatter.py:
```diff
@@ -0,0 +1,244 @@
+
+    # Mix small and regular regions
+    if mix:
+        regions = mix_small_regions(regions, target_bin_size)
```
According to codecov there is no test for mix==True. That should be added.

Redmar-van-den-Berg replied:
Added tests with mix=true

rhpvorderman commented on src/chunked_scatter/safe_scatter.py:
```diff
@@ -0,0 +1,244 @@
+    regular_regions = [reg for reg in regions if len(reg) >= target_bin_size]
+    small_regions = [reg for reg in regions if len(reg) < target_bin_size]
+
+    # We don't want to get divide by zero errors
```
```
try:
    small_regions_ratio = math.ceil(len(small_regions), len(regular_regions))
except ZeroDivisionError:
    small_regions_ratio = 1
```
Accomplishes the same thing and also prevents ZeroDivisionError in a more transparant way.

Redmar-van-den-Berg replied:
Changed

rhpvorderman commented on src/chunked_scatter/safe_scatter.py:
```diff
@@ -0,0 +1,244 @@
+    return parser
+
+
+def main():
```
You can add a simple test that tests the main function. That should improve your coverage problem while also testing the main logic (which is good to test as well).

See examples in test_main.py

Redmar-van-den-Berg replied:
Test coverage is now 100% for safe_scatter, as it should be :monocle_face:.

rhpvorderman reviewed (changes requested):
Looks good! And with nice tests as well. Some minor nitpicks and request for a test for the main function.

Redmar-van-den-Berg committed:
Add test for mixed safe scatter

Redmar-van-den-Berg committed:
Make check for zero division explicit

Redmar-van-den-Berg committed:
Fix bug where too many scatters were generated

This happens when the estimated scatter size does not cleanly divide the
total size of the specified reference. In this case, you can end up with
an extra, very small scatter.

This commit fixes this issue by keeping track of the number of generated
scatters, and adding the regions of the additional scatter to the
previous one. This way, there will always be --scatter-count regions
generated.

Redmar-van-den-Berg committed:
Add tests for the main function of safe_scatter

Redmar-van-den-Berg committed:
Fix some linting issues

Redmar-van-den-Berg committed:
Add github workflow ci file

Redmar-van-den-Berg committed:
Merge pull request #12 from biowdl/github-workflow

Add GitHub workflow

Redmar-van-den-Berg committed:
Remove travis configuration
